### PR TITLE
[multistage] Register Pinot Transform Functions into Calcite

### PR DESCRIPTION
**Pull request by xiangfu0**
- Introduce PinotSqlTransformFunction
- Use jsonExtractScalar as first example, which supports multiple(3 or 4) optional arguments.

### CODE REVIEW

File: pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/JsonPathClusterIntegrationTest.java [hunk 1]
```diff
@@ -260,15 +260,19 @@ public void testComplexQueries()
...
         + " sum(jsonExtractScalar(complexMapStr,'$.k4.met','INT'))" + " from " + DEFAULT_TABLE_NAME
         + " group by jsonExtractScalar(complexMapStr,'$.k1','STRING')"
         + " order by sum(jsonExtractScalar(complexMapStr,'$.k4.met','INT')) DESC";
-    pinotResponse = postQuery(query);
+    JsonNode pinotResponse = postQuery(query);
     Assert.assertNotNull(pinotResponse.get("resultTable").get("rows"));
-    rows = (ArrayNode) pinotResponse.get("resultTable").get("rows");
+    ArrayNode rows = (ArrayNode) pinotResponse.get("resultTable").get("rows");
     for (int i = 0; i < rows.size(); i++) {
       String seqId = _sortedSequenceIds.get(NUM_TOTAL_DOCS - 1 - i);
```
Comment (xiangfu0) [inline pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/JsonPathClusterIntegrationTest.java:277]: Actually V1 results is wrong, when it's order by `sum(jsonExtractScalar(complexMapStr,'$.k4.met','INT'))`, the result is not in order of numbers, but lex.

File: pinot-query-planner/src/main/java/org/apache/calcite/sql/fun/PinotOperatorTable.java [hunk 2]
```diff
@@ -248,4 +316,16 @@ private boolean isPinotAggregationFunction(String name) {
...
     return aggFunctionType != null && !aggFunctionType.isNativeCalciteAggregationFunctionType();
   }
+
+  private boolean isPinotTransformFunction(String name) {
+    TransformFunctionType transformFunctionType = null;
+    if (isTransformFunctionRegisteredWithOperatorTable(name)) {
+      try {
+        transformFunctionType = TransformFunctionType.valueOf(name);
+      } catch (IllegalArgumentException e) {
+        // Ignore
```
Comment (abhioncbr) [inline pinot-query-planner/src/main/java/org/apache/calcite/sql/fun/PinotOperatorTable.java:326]: Any particular reason for ignoring this exception?

**Comment by abhioncbr** [comment]
Can we add the [UTs](https://github.com/apache/pinot/blob/aee992c9a79b1651de0d37979b8c1a56bf2cea53/pinot-query-runtime/src/test/resources/queries/JsonType.json) for the jsonExtractScalar? Thanks.

Let me know; I can push the UT changes.

File: pinot-query-planner/src/main/java/org/apache/calcite/sql/fun/PinotOperatorTable.java (hunk 2 above)
Comment (xiangfu0) [reply pinot-query-planner/src/main/java/org/apache/calcite/sql/fun/PinotOperatorTable.java:326]: Expect null if function is not recognized.

**Comment by xiangfu0** [comment]
> Can we add the [UTs](https://github.com/apache/pinot/blob/aee992c9a79b1651de0d37979b8c1a56bf2cea53/pinot-query-runtime/src/test/resources/queries/JsonType.json) for the jsonExtractScalar? Thanks.
> 
> Let me know; I can push the UT changes.

I will add. Meanwhile, I will wait for 


> Can we add the [UTs](https://github.com/apache/pinot/blob/aee992c9a79b1651de0d37979b8c1a56bf2cea53/pinot-query-runtime/src/test/resources/queries/JsonType.json) for the jsonExtractScalar? Thanks.
> 
> Let me know; I can push the UT changes.

Added support for jsonExtractKey

File: pinot-common/src/main/java/org/apache/pinot/common/function/FunctionRegistry.java [hunk 3]
```diff
@@ -187,7 +187,7 @@ public static Object jsonExtractScalar(String jsonFieldName, String jsonPath, St
     }
 
     @ScalarFunction(names = {"jsonExtractKey", "json_extract_key"}, isPlaceholder = true)
-    public static String jsonExtractKey(String jsonFieldName, String jsonPath) {
+    public static Object jsonExtractKey(String jsonFieldName, String jsonPath) {
```
Comment (walterddr) [inline pinot-common/src/main/java/org/apache/pinot/common/function/FunctionRegistry.java:190]: JsonExtractKey always returns String right? a JSON key cannot be any other type?

File: pinot-common/src/main/java/org/apache/pinot/common/function/TransformFunctionType.java [hunk 4]
```diff
@@ -76,8 +89,13 @@ public enum TransformFunctionType {
   CAST("cast"),
 
   // string functions
-  JSONEXTRACTSCALAR("jsonExtractScalar"),
-  JSONEXTRACTKEY("jsonExtractKey"),
+  JSONEXTRACTSCALAR("jsonExtractScalar", ReturnTypes.cascade(
+      opBinding -> inferJsonExtractScalarExplicitTypeSpec(opBinding).orElse(
+          opBinding.getTypeFactory().createSqlType(SqlTypeName.VARCHAR, 2000)), SqlTypeTransforms.FORCE_NULLABLE),
```
Comment (walterddr) [inline pinot-common/src/main/java/org/apache/pinot/common/function/TransformFunctionType.java:94]: we dont need to limit varchar to 2000 here

File: pinot-common/src/main/java/org/apache/pinot/common/function/TransformFunctionType.java [hunk 5]
```diff
@@ -180,4 +210,46 @@ public String getName() {
...
+  private static Optional<RelDataType> inferJsonExtractScalarExplicitTypeSpec(SqlOperatorBinding opBinding) {
+    if (opBinding.getOperandCount() > 2
+        && opBinding.isOperandLiteral(2, false)) {
+      String operandType = opBinding.getOperandLiteralValue(2, String.class).toUpperCase();
+      return Optional.of(inferExplicitTypeSpec(operandType, opBinding.getTypeFactory()));
+    }
+    return Optional.empty();
+  }
+
+  private static RelDataType inferExplicitTypeSpec(String operandType, RelDataTypeFactory typeFactory) {
```
Comment (walterddr) [inline pinot-common/src/main/java/org/apache/pinot/common/function/TransformFunctionType.java:232]:
```suggestion
  private static RelDataType inferExplicitTypeSpec(String operandTypeStr, RelDataTypeFactory typeFactory) {
```

File: pinot-common/src/main/java/org/apache/pinot/common/function/TransformFunctionType.java [hunk 6]
```diff
@@ -180,4 +210,46 @@ public String getName() {
...
+    switch (operandType) {
+      case "INT":
+      case "INTEGER":
```
Comment (walterddr) [inline pinot-common/src/main/java/org/apache/pinot/common/function/TransformFunctionType.java:235]: INTEGER is valid SQL type name

File: pinot-common/src/main/java/org/apache/pinot/common/utils/DataSchema.java [hunk 7]
```diff
@@ -298,24 +298,32 @@ public boolean isSuperTypeOf(ColumnDataType subTypeCandidate) {
     public DataType toDataType() {
       switch (this) {
         case INT:
+        case INT_ARRAY:
```
Comment (walterddr) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/DataSchema.java:301]: why this change? let's dont mix array support in here if we can make that separate

File: pinot-query-planner/src/main/java/org/apache/calcite/sql/fun/PinotOperatorTable.java [hunk 8]
```diff
@@ -207,6 +222,49 @@ public final void initNoDuplicate() {
     }
   }
 
+  /**
+   * This registers transform functions defined in {@link org.apache.pinot.common.function.TransformFunctionType}
+   * which are multistage enabled.
+   */
+  private void initTransformFunctions() {
```
Comment (walterddr) [inline pinot-query-planner/src/main/java/org/apache/calcite/sql/fun/PinotOperatorTable.java:229]: see my change in https://github.com/apache/pinot/pull/11068 
this method can be greatly simplified

File: pinot-query-planner/src/main/java/org/apache/calcite/sql/fun/PinotOperatorTable.java (hunk 2 above)
Comment (walterddr) [reply pinot-query-planner/src/main/java/org/apache/calcite/sql/fun/PinotOperatorTable.java:326]: this error shouldn't occur if the isTransform... method is correctly capture the name

File: pinot-common/src/main/java/org/apache/pinot/common/function/FunctionRegistry.java (hunk 3 above)
Comment (xiangfu0) [reply pinot-common/src/main/java/org/apache/pinot/common/function/FunctionRegistry.java:190]: From the test, seem the result type is an array of string

File: pinot-common/src/main/java/org/apache/pinot/common/utils/DataSchema.java (hunk 7 above)
Comment (xiangfu0) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/DataSchema.java:301]: The issue is that json_extract_key returns array.

File: pinot-common/src/main/java/org/apache/pinot/common/function/FunctionRegistry.java (hunk 3 above)
Comment (walterddr) [reply pinot-common/src/main/java/org/apache/pinot/common/function/FunctionRegistry.java:190]: ah.... i see

File: pinot-common/src/main/java/org/apache/pinot/common/utils/DataSchema.java (hunk 7 above)
Comment (walterddr) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/DataSchema.java:301]: i see. not sure if there's going to be any side effect of this change going forward with ARRAY supports. 
can you add support like
`SELECT cardinality(JSONEXTRACTKEY(...))` and see if that works?

File: pinot-query-planner/src/main/java/org/apache/calcite/sql/fun/PinotOperatorTable.java (hunk 8 above)
Comment (xiangfu0) [reply pinot-query-planner/src/main/java/org/apache/calcite/sql/fun/PinotOperatorTable.java:229]: ack

File: pinot-common/src/main/java/org/apache/pinot/common/utils/DataSchema.java (hunk 7 above)
Comment (xiangfu0) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/DataSchema.java:301]: Added a sample query in unit test.

File: pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/MultiStageEngineJsonPathClusterIntegrationTest.java [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+import com.fasterxml.jackson.databind.node.ArrayNode;
+import com.google.common.collect.ImmutableMap;
+import java.util.Properties;
+import org.apache.pinot.client.Connection;
+import org.apache.pinot.client.ConnectionFactory;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+
+public class MultiStageEngineJsonPathClusterIntegrationTest extends JsonPathClusterIntegrationTest {
```
Comment (xiangfu0) [inline pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/MultiStageEngineJsonPathClusterIntegrationTest.java:31]: This test can be simplified once https://github.com/apache/pinot/pull/11064 is merged.

File: pinot-query-planner/src/main/java/org/apache/calcite/sql/fun/PinotOperatorTable.java [hunk 10]
```diff
@@ -207,6 +222,49 @@ public final void initNoDuplicate() {
...
+    // Walk through all the Pinot transform types and register those that are supported in multistage.
+    for (TransformFunctionType transformFunctionType : TransformFunctionType.values()) {
+      if (transformFunctionType.getSqlKind() == null) {
+        // Skip registering functions which are standard Calcite functions and functions which are not yet supported.
+        continue;
+      }
+
+      // Register the aggregation function with Calcite along with all alternative names
+      List<PinotSqlTransformFunction> sqlTransformFunctions = new ArrayList<>();
```
Comment (walterddr) [inline pinot-query-planner/src/main/java/org/apache/calcite/sql/fun/PinotOperatorTable.java:238]: this list is not necessary, you can simply register this one-by-one

File: pinot-query-planner/src/main/java/org/apache/calcite/sql/fun/PinotOperatorTable.java [hunk 11]
```diff
@@ -207,6 +222,49 @@ public final void initNoDuplicate() {
...
+      PinotSqlTransformFunction aggFunction = generatePinotSqlTransformFunction(transformFunctionType.getName(),
+          transformFunctionType, false);
+      sqlTransformFunctions.add(aggFunction);
+      List<String> alternativeFunctionNames = transformFunctionType.getAliases();
+      if (alternativeFunctionNames == null || alternativeFunctionNames.size() == 0) {
+        // If no alternative function names are specified, generate one which converts camel case to have underscores
+        // as delimiters instead. E.g. boolAnd -> BOOL_AND
+        String alternativeFunctionName =
+            convertCamelCaseToUseUnderscores(transformFunctionType.getName());
```
Comment (walterddr) [inline pinot-query-planner/src/main/java/org/apache/calcite/sql/fun/PinotOperatorTable.java:247]: this convert should be moved to transform function type. this is the way we going to handle `_` separated vs camel case unless specific naming is specified.

File: pinot-query-planner/src/main/java/org/apache/calcite/sql/fun/PinotOperatorTable.java (hunk 10 above)
Comment (xiangfu0) [reply pinot-query-planner/src/main/java/org/apache/calcite/sql/fun/PinotOperatorTable.java]: done.

File: pinot-query-planner/src/main/java/org/apache/calcite/sql/fun/PinotOperatorTable.java (hunk 11 above)
Comment (xiangfu0) [reply pinot-query-planner/src/main/java/org/apache/calcite/sql/fun/PinotOperatorTable.java]: removed.

**Review by walterddr — APPROVED** [review]
lgtm
